src/main/scala/firrtl/transforms/InlineCasts.scala

```diff
@@ -28,23 +28,30 @@ object InlineCastsTransform {
     * @param expr the Expression being transformed
     * @return Returns expr with [[WRef]]s replaced by values found in replace
     */
-  def onExpr(replace: NodeMap)(expr: Expression): Expression = expr match {
-    // Anything that may generate a part-select should not be inlined!
-    case DoPrim(op, _, _, _) if (isBitExtract(op) || op == Pad) => expr
-    case e =>
-      e.map(onExpr(replace)) match {
-        case e @ WRef(name, _, _, _) =>
-          replace
-            .get(name)
-            .filter(isSimpleCast(castSeen = false))
-            .getOrElse(e)
-        case e @ DoPrim(op, Seq(WRef(name, _, _, _)), _, _) if isCast(op) =>
-          replace
-            .get(name)
-            .map(value => e.copy(args = Seq(value)))
-            .getOrElse(e)
-        case other => other // Not a candidate
-      }
+  def onExpr(replace: NodeMap)(expr: Expression): Expression = {
+    // Keep track if we've seen any non-cast expressions while recursing
+    def rec(seenComplexExpr: Boolean)(expr: Expression): Expression = expr match {
```

Comment:
```suggestion
    def rec(hasNonCastParent: Boolean)(expr: Expression): Expression = expr match {
```

Directly referencing what this means for the structure of the tree might make it clearer.

Reply:
Keeping the "complex" is fine, I just think that "has<X>Parent" is clearer than "seen."

Reply:
`hasNonCastParent` is better I think, will change.